### PR TITLE
Updates and tweaks which enable cross-building of 64bit arm binaries

### PR DESCRIPTION
**Pull request by miodragpop**
cross-build process is still several times faster than native build on Raspberry Pi 4 ...

`sudo apt-get install gcc-aarch64-linux-gnu g++-aarch64-linux-gnu`
`zcutil/build-cross-aarch64.sh -j$(nproc)`

### CODE REVIEW

File: depends/packages/proton.mk [hunk 1]
```diff
@@ -11,7 +11,7 @@ define $(package)_preprocess_cmds
 endef
 
 define $(package)_config_cmds
-  cd build; cmake .. -DCMAKE_CXX_STANDARD=11 -DCMAKE_INSTALL_PREFIX=/ -DSYSINSTALL_BINDINGS=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DBUILD_PYTHON=OFF -DBUILD_PHP=OFF -DBUILD_JAVA=OFF -DBUILD_PERL=OFF -DBUILD_RUBY=OFF -DBUILD_JAVASCRIPT=OFF -DBUILD_GO=OFF -DBUILD_STATIC_LIBS=ON
+  cd build; cmake .. -DCMAKE_CXX_STANDARD=11 -DCMAKE_INSTALL_PREFIX=/ -DSYSINSTALL_BINDINGS=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DBUILD_PYTHON=OFF -DBUILD_RUBY=OFF -DBUILD_GO=OFF -DBUILD_STATIC_LIBS=ON -DLIB_SUFFIX= -DENABLE_JSONCPP=
```
Comment (leto) [inline depends/packages/proton.mk:14]: Why are these changes necessary here? I am not against updating proton version, but why are compile flags changing? Did this come from KMD or BTC or ZEC?

File: depends/packages/rust.mk [hunk 2]
```diff
@@ -1,28 +1,33 @@
 package=rust
-$(package)_version=1.32.0
+$(package)_version=1.42.0
```
Comment (leto) [inline depends/packages/rust.mk:2]: Is this necessary or optional? It's a big change that will require testing many librustzcash codepaths, that only our lightwalletd server may use

File: depends/packages/proton.mk (hunk 1 above)
Comment (miodragpop) [reply depends/packages/proton.mk:14]: I've replaced complete file, haven't edited line by line. Mostly it's order of flags that's changed, not the flags themselves. Taken from zcash depends.

File: depends/packages/rust.mk (hunk 2 above)
Comment (miodragpop) [reply depends/packages/rust.mk:2]: The same story, taken from zcash depends. Because it crossbuilds without issues, no other reason.

File: src/cryptoconditions/src/include/secp256k1/Makefile.am [hunk 3]
```diff
@@ -71,7 +71,7 @@ endif
 endif
 
 libsecp256k1_la_SOURCES = src/secp256k1.c
-libsecp256k1_la_CPPFLAGS = -DSECP256K1_BUILD -I$(top_srcdir)/include -I$(top_srcdir)/src $(SECP_INCLUDES) -march=x86-64 -g
+libsecp256k1_la_CPPFLAGS = -DSECP256K1_BUILD -I$(top_srcdir)/include -I$(top_srcdir)/src $(SECP_INCLUDES) -g
```
Comment (leto) [inline src/cryptoconditions/src/include/secp256k1/Makefile.am:74]: This looks like a good bugfix, this always assumed x86 64bit

File: depends/packages/rust.mk (hunk 2 above)
Comment (leto) [reply depends/packages/rust.mk:2]: this could break things, we should make upgrading Rust a different PR that can be tested by itself. Many crates depend on Rust version, and even if hushd works perfectly, something in lightwalletd could break. @DenioD you have comments?

**Comment by leto** [comment]
@miodragpop please delete Proton completely from Hush as part of your PR, which should simplify it

File: src/cc/Makefile_custom [hunk 4]
```diff
@@ -2,6 +2,7 @@ SHELL		= /bin/sh
 CC		= gcc
 CC_DARWIN	= g++-8
 CC_WIN		= x86_64-w64-mingw32-gcc-posix
+CC_ARM64	= aarch64-linux-gnu-g++
```
Comment (leto) [inline src/cc/Makefile_custom:5]: should this be called CC_ARM64 or CC_AARCH64 ?

File: src/cc/Makefile_custom [hunk 5]
```diff
@@ -28,6 +29,10 @@ else ifeq ($(HOST),x86_64-w64-mingw32)
 	$(CC_WIN) $(CFLAGS_WIN) $(DEBUGFLAGS) -o $(TARGET_WIN) -c $(SOURCES)
 	cp $(TARGET_WIN) ../libcc.dll
 #else ifeq ($(WIN_HOST),True) - todo: pass ENV var from build.sh if WIN host
+else ifeq ($(HOST),aarch64-linux-gnu)
```
Comment (leto) [inline src/cc/Makefile_custom:32]: These and any other CC  changes may be desired by KMD upstream, fyi

**Comment by leto** [comment]
@miodragpop if you have some unnecessary whitespace changes (which I think I see in various places), I prefer not to see those, unless it's fixing some bad formatting. If so, whitespace changes should be made in their own commit and not mixed with functional changes, going forward

**Comment by leto** [comment]
@ssadler some of this PR changes the CC build system, that looks like it should be sent to KMD. I was wondering if you could code review and give comments about sending it upstream

File: depends/packages/proton.mk (hunk 1 above)
Comment (leto) [reply depends/packages/proton.mk:14]: `-DBUILD_JAVA=OFF -DBUILD_PERL=OFF -DBUILD_PHP=OFF` was lost, but we are deleting Proton, so it doesn't matter 🎉

**Comment by ssadler** [comment]
> @ssadler some of this PR changes the CC build system, that looks like it should be sent to KMD. I was wondering if you could code review and give comments about sending it upstream

Hi, it appears to alter the CC build system but in practice it does not; secp256k1 inside the cryptoconditions tree is not part of the `cryptoconditions/libcryptoconditions_core.la` build target used by komodo.

The change made to the secp256k1 target removes the `-march` parameter which was added for verus in https://github.com/jl777/komodo/commit/f2c103207f5137029bffd2c948d202498433eb89, but build configs aren't my specialty and I have no idea why! I think improvements for building on ARM will be welcome upstream though.

File: depends/packages/proton.mk (hunk 1 above)
Comment (miodragpop) [reply depends/packages/proton.mk]: proton removed

File: depends/packages/rust.mk (hunk 2 above)
Comment (miodragpop) [reply depends/packages/rust.mk]: rust reverted to 1.32.0

**Comment by miodragpop** [comment]
build script for aarch64 deleted ... cross build is available via default build.sh as in `HOST=aarch64-linux-gnu zcutil/build.sh`

**Comment by leto** [comment]
@miodragpop that is a huge improvement! Impressive

File: depends/packages/libcurl.mk [hunk 6]
```diff
@@ -4,7 +4,7 @@ $(package)_dependencies=openssl
 $(package)_download_path=https://curl.haxx.se/download
 $(package)_file_name=curl-$($(package)_version).tar.gz
 $(package)_sha256_hash=52af3361cf806330b88b4fe6f483b6844209d47ae196ac46da4de59bb361ab02
-$(package)_config_opts_linux=--disable-shared --enable-static --prefix=$(host_prefix) --host=x86_64-unknown-linux-gnu
+$(package)_config_opts_linux=--disable-shared --enable-static --prefix=$(host_prefix) --host=$(host)
 $(package)_config_opts_mingw32=--enable-mingw --disable-shared --enable-static --prefix=$(host_prefix) --host=x86_64-w64-mingw32
```
Comment (leto) [inline depends/packages/libcurl.mk:8]: Does this line also need a `--host=$(host)` ?

File: depends/packages/packages.mk [hunk 7]
```diff
@@ -1,12 +1,4 @@
-rust_packages := rust librustzcash
-
-ifeq ($(build_os),darwin)
-	zcash_packages :=  libgmp libsodium utfcpp
-else
-	proton_packages := proton
-	zcash_packages := libgmp libsodium utfcpp
-endif
-
+zcash_packages :=  libgmp libsodium utfcpp
```
Comment (leto) [inline depends/packages/packages.mk:1]: This diff makes me so happy

File: depends/packages/rust.mk [hunk 8]
```diff
@@ -1,28 +1,33 @@
...
-endif
+# Mapping from GCC canonical hosts to Rust targets
+# If a mapping is not present, we assume they are identical, unless $host_os is
+# "darwin", in which case we assume x86_64-apple-darwin.
+$(package)_rust_target_x86_64-w64-mingw32=x86_64-pc-windows-gnu
+
+# Mapping from Rust targets to SHA-256 hashes
+$(package)_rust_std_sha256_hash_aarch64-unknown-linux-gnu=346efe3aef2aff7b71a611bf7661bcec5f9bc4025a599c2866ec5fd330247cb9
+$(package)_rust_std_sha256_hash_x86_64-apple-darwin=b736d035a97f830585360e54e3f8877b68c942211cf0a75e805f34bfb36103a6
+$(package)_rust_std_sha256_hash_x86_64-pc-windows-gnu=cad5f1454d591c13eeb3657f1c9dbfeb30e648f59680bd0765b94c63e7afc49e
```
Comment (leto) [inline depends/packages/rust.mk:19]: Can you comment on exactly which files these are the hashes for?

File: src/cc/makecustom [hunk 9]
```diff
@@ -1,5 +1,5 @@
 #!/bin/sh
-if make -f Makefile_custom "$@"; then
+if HOST="$HOST" make -B -f Makefile_custom "$@"; then
```
Comment (leto) [inline src/cc/makecustom:2]: This is a nice diff that KMD upstream will probably want as well

File: src/cryptoconditions/src/asn/OCTET_STRING.c [hunk 10]
```diff
@@ -1731,6 +1733,15 @@ OCTET_STRING_free(asn_TYPE_descriptor_t *td, void *sptr, int contents_only) {
...
 
+/* Attention !!!
+ * this is quick & dirty workaround for memory corruption bug on aarch64-linux-gnu
+ * - downside: allows memory leakage
+ * - issue description: On Raspberry Pi 4 @ 64bit linux, daemon crashes with "free(): invalid pointer" error
+ * - probable cause: misaligned memory access to nested structs containing pointers
+ * - TODO: use the latest asn1c compiler on CryptoConditions.asn, maybe generate cpp instead of c code... investigation in progress
+ */
+
+#if !defined(__aarch64__)
```
Comment (leto) [inline src/cryptoconditions/src/asn/OCTET_STRING.c:1744]: This was a very smart workaround to let you sync the chain and test other stuff. I think the best way long-term is that I explicitly disable all known CCs on HUSH3, but leave open the possibility of zaddr CC's in the future, because we have our own custom "Hush CC" we can customize and allow it to make special taddr tx's in the future. Just like KMD notaries can make dpow tx's. 

New Hush Smart Chains can choose to enable or disable CC's (they have always defaulted to off), and when we have some kind of zCC in the future, we can push it to mainnet when it's ready 😄

Comment (leto) [reply src/cryptoconditions/src/asn/OCTET_STRING.c:1744]: We can probably leave this workaround for now, so you can sync, and once my code that ignores CC's is done, this code should change to always define the stack, but if NAME=HUSH3 and arch=aarch64, just leak memory instead of crashing. Until we get this upstream CC bug fixed on aarch64

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -2439,11 +2439,10 @@ void CWallet::DeleteTransactions(std::vector<uint256> &removeTxs) {
             return;
         }
     }
-#if defined(__unix__) || defined(_WIN64)
-    // Miodrag: release memory back to the OS
+#if defined(__GLIBC__)
     malloc_trim(0);
 #else
-    //TODO: This doesn't work on Mac
+    // On Mac and Win memory isn't kept back upon vector or list member erase, different garbage collector strategy. No need to force trimming.
```
Comment (leto) [inline src/wallet/wallet.cpp:2445]: This useful comment will save countless debugging hours in the future, thank you
